### PR TITLE
TMOP combined with VM. [tmop-dev]

### PR DESCRIPTION
**Pull request by vladotomov**
Some TMOP-based additional HyperelasticModels and code to handle them in HyperelasticNLFIntegrator. I also added some comments in HyperElasticModel, but I'm not sure if they’re relevant for other use cases.

Companion documentation PR, here: https://github.com/mfem/web/pull/7

### TODO
- [x] Update CHANGELOG
- [x] Add a simple mesh optimization example in the miniapps/meshing sub-directory
- [x] Consider moving new mesh files to the [mfem/data repo](https://github.com/mfem/data)
- [x] Consider renaming `TMOP` classes to something does not involving `elastic`
- [x] Update `.gitignore`

### After merging
- [ ] Consider removing the Mesquite support
- [x] Update web documentation: https://github.com/mfem/web/pull/7

### CODE REVIEW

**Comment by tzanio** [comment]
After this is merged, I suggest that we consider removing the Mesquite support...

**Comment by tzanio** [comment]
The `ball-pert.mesh` is quite big. I suggest we move it to the external [mfem/data repo](https://github.com/mfem/data).

**Comment by tzanio** [comment]
For the rest of the mesh files -- I suggest we move them either to `miniapps/meshing` or also to the external `mfem/data` repo.

File: fem/nonlininteg.cpp [hunk 1]
```diff
@@ -247,114 +448,1655 @@ void NeoHookeanModel::AssembleH(const DenseMatrix &J, const DenseMatrix &DS,
             }
 }
 
+double TMOPHyperelasticModel001::EvalW(const DenseMatrix &Jpt) const
```
Comment (tzanio) [inline fem/nonlininteg.cpp]: My preference will be to change the name as TMOP models doesn't have to be `elastic`.

**Comment by tzanio** [comment]
@Xknupp, do you mind taking a look at this initial support for TMOP in MFEM and letting us know if you have any comments or suggestions for improvements?

Differences are here: https://github.com/mfem/mfem/pull/193/files. See in particular the `CHANGELOG` file and the `Mesh Optimizer` example.

Thanks!
Tzanio

**Comment by tzanio** [comment]
@kmittal2, @v-dobrev please review if you can. It will be nice to have this approved for the PI meeting next week :)

File: fem/nonlininteg.hpp [hunk 2]
```diff
@@ -127,30 +157,49 @@ class NeoHookeanModel : public HyperelasticModel
...
+    target->physical coordinates transformation. The target configuration is
+    given by the current mesh at the time of the evaluation of the intergrator.
+*/
 class HyperelasticNLFIntegrator : public NonlinearFormIntegrator
 {
 private:
    HyperelasticModel *model;
 
-   DenseMatrix DSh, DS, J0i, J1, J, P, PMatI, PMatO;
+   //   Jrt: the Jacobian of the target-to-reference-element transformation.
```
Comment (tzanio) [inline fem/nonlininteg.hpp]: Excellent documentation!

File: linalg/invariants.hpp [hunk 3]
```diff
@@ -0,0 +1,1347 @@
...
+   }
+
+   // Assemble operation for tensor X with components X_jslt:
+   //    A(i+nd*j,k+nd*l) += (\sum_st  w D_is X_jslt D_kt)
+   //    0 <= i,k < nd,  0 <= j,l,s,t < 2
+   // where nd is the height of D, i.e. the number of DOFs in one component.
+
+   void Assemble_ddI1(scalar_t w, scalar_t *A)
+   {
+      // ddI1_jslt = 2 I_jslt = 2 δ_jl δ_st
```
Comment (tzanio) [inline linalg/invariants.hpp]: Don't think we've had Unicode characters in the code before... Are we sure we want to start now? (cc: @camierjs)

File: fem/tmop.hpp [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+
+/** @brief Base class representing target-matrix construction algorithms for
+    mesh optimization via the target-matrix optimization paradigm (TMOP). */
+/** This class is used by class TMOP_Integrator to construct the target Jacobian
+    matrices (reference-element to target-element) at quadrature points. It
+    supports a set of algorithms chosen by the #TargetType enumeration.
+
+    New target-matrix construction algorithms can be defined by deriving new
+    classes and overriding the method ComputeElementTargets(). */
+class TargetConstructor
```
Comment (tzanio) [inline fem/tmop.hpp]: Like the name

File: miniapps/meshing/CMakeLists.txt [hunk 5]
```diff
@@ -25,4 +25,24 @@ add_mfem_miniapp(shaper
   MAIN shaper.cpp
   LIBRARIES mfem)
 
-# Add tests?
```
Comment (tzanio) [inline miniapps/meshing/CMakeLists.txt]: 👍

File: makefile [hunk 6]
```diff
@@ -466,5 +468,15 @@ style:
 	fi
 
 # Print the contents of a makefile variable, e.g.: 'make print-MFEM_LIBS'.
-print-%: ; @printf "%s:\n" $*
-	@printf "%s\n" $($*)
+print-%:
```
Comment (tzanio) [inline makefile]: This is nice

File: linalg/invariants.hpp [hunk 7]
```diff
@@ -0,0 +1,1347 @@
...
+                  A[il+ah*kj] += A_ik_lj;
+                  A[kj+ah*il] += A_ik_lj;
+               }
+            }
+         }
+      }
+   }
+   void Assemble_ddI2(scalar_t w, scalar_t *A)
+   {
+      // dI2 = 2 I_1 J - 2 J J^t J = 2 (I_1 I - B) J
```
Comment (tzanio) [inline linalg/invariants.hpp]: This whole file is a really impressive mathematical tour de force!

File: fem/tmop.cpp [hunk 8]
```diff
@@ -0,0 +1,2283 @@
...
+   return Dim2Invariant1(Jpt) * Dim2Invariant2(Jpt);
+#else
+   ie.SetJacobian(Jpt.GetData());
+   return ie.Get_I1();
+#endif
+}
+
+void TMOP_Metric_001::EvalP(const DenseMatrix &Jpt, DenseMatrix &P) const
+{
+#ifdef MFEM_USE_OLD_TMOP_INVARIANTS
```
Comment (tzanio) [inline fem/tmop.cpp]: Consider removing these?

Comment (v-dobrev) [reply fem/tmop.cpp]: Yes, I'll remove the old implementation.

File: linalg/invariants.hpp (hunk 3 above)
Comment (v-dobrev) [reply linalg/invariants.hpp]: There is a string containing `"ΔTE"` in `ex10{,p}.cpp`. Also, in the comments in for class SidreDataCollection, there are some characters that are not standard ascii, not sure if those are unicode or just extended ascii. Anyway, I think we have a precedent already. ;-)

**Pull request merged**